### ISSUE TITLE
Computing gradients of predictions

### ISSUE THREAD

**Issue by marpulli** [cornellius-gp/gpytorch#416]
Hi, I'm wanting to compute the gradients of the predictive mean and variance with respect to the input location. Is this possible with the current version of gpytorch? Thanks!

**Comment by Balandat** [comment]
Yes that's possible. If you have a fitted model `model` in eval mode, and `X` is a tensor that requires a gradient, then you can do 

```
mvn = model(X)
o = f(mvn.mean, mvn.variance)
o.backward()
```

where `f` is some function of mean and variance that supports backward passes.

**Comment by marpulli** [comment]
Thanks for the reply, I've been trying something like that but my gradient is `None`. As a test I'm just summing the mean values, here is the code:
```
model.eval()
likelihood.eval()

test_x = torch.ones((1,1), requires_grad=True)
prediction = likelihood(model(test_x))
y = prediction.mean.sum()
y.backward()
print(y.grad)
```

Any ideas where I'm going wrong?

**Comment by Balandat** [comment]
Yes, since you want the gradient w.r.t. `test_x`, instead of `y.grad` you'll want to call `test_x.grad`. 

Feel free to re-open this issue if you run into any further problems with that.

**Comment by marpulli** [comment]
Ah yes, of course!! Thanks for the help

**Comment by HengyuLi-Ozaki-lab** [comment]
Sorry, did anyone solved this problem?

**Comment by jejjohnson** [comment]
I have a [demo colab notebook](https://colab.research.google.com/drive/1Byv0pLMuSFsqKtzHFaBf-8iR0PRJi7Es?usp=sharing) that demonstrates how one can calculate the gradient with respect to the inputs. This is a simple 1D demo where you can clearly see the 1st and 2nd derivative of the GP predictive mean. I do a small ad-hoc test for the Hessian (check if the diagonals are the same as the 2nd derivative). It seems to work well for my application (input uncertainty propagation).

Note: I'm a fan of the `torch.autograd.function` API as it's a bit easier with less boilerplate code.

**Comment by HengyuLi-Ozaki-lab** [comment]
> I have a [demo colab notebook](https://colab.research.google.com/drive/1Byv0pLMuSFsqKtzHFaBf-8iR0PRJi7Es?usp=sharing) that demonstrates how one can calculate the gradient with respect to the inputs. This is a simple 1D demo where you can clearly see the 1st and 2nd derivative of the GP predictive mean. I do a small ad-hoc test for the Hessian (check if the diagonals are the same as the 2nd derivative). It seems to work well for my application (input error propagation).
> 
> Note: I'm a fan of the `torch.autograd.function` API as it's a bit easier with less boilerplate code.

Wow, that‘s awesome, thanks a lot!
